### PR TITLE
Implement unstable `-Clinker-flavor=gcc:lld` for MCP 510

### PR DESCRIPTION
**Pull request by lqd**
This PR is extracted from #96401 for easier review.

It implements half of [MCP #510](https://github.com/rust-lang/compiler-team/issues/510): adding a `gcc:lld` linker-flavor on the CLI, as a prerequisite for supporting `rust-lld` directly via a command similar to `-Clinker-flavor=gcc:lld -Clink-self-contained=linker`.

The option is unstable and requires `-Z unstable-options` to be used.

When used without `-Clink-self-contained=linker`, it's equivalent to `-Clink-arg=-fuse-ld=lld`.

r? @petrochenkov

### CODE REVIEW

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 1]
```diff
@@ -2693,46 +2699,126 @@ fn get_apple_sdk_root(sdk_name: &str) -> Result<String, String> {
...
+    // adding an argument to the `cc` invocation to use the `use_ld` given linker.
+    let use_ld = match &cg.linker_flavor {
+        Some(LinkerFlavorCli::Gcc { use_ld }) => {
+            // Ensure `gcc` is the currently selected flavor. Error out cleanly, as `-Zgcc-ld` does
+            // if that happens, but this should be unreachable.
+            if LinkerFlavor::Gcc != flavor {
+                sess.fatal(
+                    "`-Clinker-flavor=gcc:*` flag is used even though the \
+                    linker flavor is not `gcc`",
+                );
```
Comment (bjorn3) [inline compiler/rustc_codegen_ssa/src/back/link.rs:2733]: How can this happen? They should always be in sync by definition, right?

Comment (lqd) [reply compiler/rustc_codegen_ssa/src/back/link.rs:2731]: It's not supposed to indeed, as long as this piece of code stays downstream from the CLI linker-flavor lowering, and the actual `LinkerFlavor` is not changed or overriden by the time the order-independent options are added to the cmd args. If this happens by mistake (or intentionally) in the future, this check will catch it, but it's not going to trigger today.

Comment (bjorn3) [reply compiler/rustc_codegen_ssa/src/back/link.rs:2731]: Maybe change it to a `bug!()` instead?

Comment (lqd) [reply compiler/rustc_codegen_ssa/src/back/link.rs:2736]: Good idea Thank you.

I've pushed a commit to address this.

**Comment by petrochenkov** [comment]
I suddenly got busy and will get to this PR (and then to https://github.com/rust-lang/rust/pull/96884) next weekend most likely.

**Comment by petrochenkov** [comment]
>`gcc` linker-flavor accept a colon-separator to detect what to pass to `cc`. For example, `-Clinker-flavor=gcc:$linker` will be passed as `-fuse-ld=$linker` verbatim.

But that's not a linker flavor, that's literally `-Zgcc-ld` that you (we?) have been trying to eliminate.

If that was a linker flavor, then we'd add
```
"gcc-ld"
"gcc-wasm-ld"
"gcc-ld64.lld"
"gcc-ld.lld"
"gcc-lld-link"
```
to already existing
```
"ld"
"wasm-ld"
"ld64.lld"
"ld.lld"
"lld-link"
```
flavors.
No strings passed around verbatim, only "generalized classes" of linkers.

If we want to pass strings to `-fuse-ld` verbatim then it's better to keep a separate option like `-Zgcc-ld`.
At the same time, half of the concerns on the Zulip thread was about how "passing strings to linker is bad and we need something more high level".

**Comment by petrochenkov** [comment]
(I need to reread the zulip thread because I'm starting to forget all of the colors of this shed.)

**Comment by petrochenkov** [comment]
Sorry for the delay.
I'm now back from vacation and will return to this soon.

**Comment by petrochenkov** [comment]
Update: I'm still against having a free form component in `-Clinker-flavor`, but I don't have time to write a good motivation and finish my experiments in https://github.com/petrochenkov/rust/commit/836063a683f0122d2c59e03fffcf4eaac7a321b2 / https://github.com/petrochenkov/rust/commit/2227d27dece96339dc1c9d31b21b1fa0ec23c551.
The timing has been really unlucky here, once I'm less busy and can allocate a few continuous hours of Rust work, this is the first thing I'm going to address.

**Comment by lqd** [comment]
Rebased.

> Update: I'm still against having a free form component in `-Clinker-flavor`

@petrochenkov I've pushed a separate commit to remove that, and use the existing `LdImpl` enum instead. 

The two experiments are interesting but look more like general refactorings that don't necessarily impact the CLI for the MCP or prevent from making progress here until stabilization ? (and when they do impact the CLI, this can be seen as breaking changes on stable flags, something that I've avoided)

**Comment by petrochenkov** [comment]
Sorry for the delay.

So I wanted to try documenting what "linker flavor" even means.
(I'll consider both `-Clinker-flavor` and target specs.)

All linkers have some kind of command line interfaces and rustc needs to know which commands to use with each of them.
So we clusterize all such interfaces into a (somewhat arbitrary) number of classes and call them "linker flavors".

Technically, it's not even necessary, we can nearly always infer the flavor from linker name and target properties like `is_like_windows`/`is_like_osx`/etc.
However, the PRs originally introducing `-Zlinker-flavor` (https://github.com/rust-lang/rust/pull/40018 and friends) were aiming to reduce this kind of inference and provide something certain and explicitly specified instead.

The minimal set of linker flavors could look something like this `(1)`
```rust
// Unix-like linker with GNU extensions (both naked and compiler-wrapped forms)
// Besides similar "default" Linux/BSD linkers this also include Windows/GNU, which is somewhat different.
Gnu { cc: bool },
// Basic Unix-like linker, possibly with non-GNU extensions aka "any other Unix" (both naked and compiler-wrapped forms)
// Apple targets, Wasm, Solaris/illumos, l4re, msp430 - very different targets.
// Why this exists separately from GNU - the `linker_is_gnu` target property (which I will include into linker flavor in this post)
// is annoying to infer using other target properties.
Unix { cc: bool },
// MSVC linker has a unique interface
Msvc,
// Other linker-like tools with unique interfaces for exotic targets
EmCc,
Bpf,
Ptx,
```

Then LLD appears and real life forces our hand to increase the number of clusters to be a bit less minimal.
LLD linker is self-sufficent only if its executable (`argv[0]`) is named in specific way, otherwise it doesn't work and requires a `-flavor LLD_FLAVOR` argument to choose which logic to use.
Our shipped `rust-lld` in particular is *not* named in such specific way, maybe it was a mistake to ship it like this, but it's too late now.
So it needs the flavor option, so let's also make it "certain and explicit" and not inferred from other target properties.
(I'll also assume we want to call LLD through a C compiler too, which is the point of this PR.)

This changes our set of linker flavors to something like this `(2)`:
```rust
// Unix-like linker with GNU extensions (both naked and compiler-wrapped forms)
// Besides similar "default" Linux/BSD linkers this also include Windows/GNU, which is somewhat different.
Gnu { lld: bool, cc: bool },
// Unix-like linker for Apple targets (both naked and compiler-wrapped forms)
// Why it was extracted from "umbrella" Unix - due to the corresponding LLD flavor.
Darwin { lld: bool, cc: bool },
// Unix-like linker for Wasm targets (both naked and compiler-wrapped forms)
// Why it was extracted from "umbrella" Unix - due to the corresponding LLD flavor.
// Non-LLD version does not exist, so the lld flag here is currently hardcoded and not boolean.
WasmLld { cc: bool },
// Basic Unix-like linker, possibly with non-GNU extensions aka "any other Unix" (both naked and compiler-wrapped forms)
// Solaris/illumos, l4re, msp430 - very different targets.
// LLD doesn't support any of these.
Unix { cc: bool },
// MSVC linker has a unique interface
// LLD supports it
Msvc { lld: bool },
// Other linker-like tools with unique interfaces for exotic targets
EmCc,
Bpf,
Ptx,
```

Then we could increase the number of clusters even more, e.g. to cover most of conditions in `compiler\rustc_codegen_ssa\src\back\linker.rs` `(3)`:
```rust
Gnu { lld: bool, cc: bool },
Mingw { lld: bool, cc: bool }, // split from GNU
Darwin { lld: bool, cc: bool },
WasmLld { cc: bool },
Solaris { cc: bool }, // split from Unix
L4Re { cc: bool }, // split from Unix
Unix { cc: bool }, // "other Unix" still needs to exist
Msvc { lld: bool },
EmCc,
Bpf,
Ptx,
```
We can realistically do this, but it may be too much for my taste. It would *still* not eliminate all the target-specific logic in the rustc_codegen_ssa though.
Anyway, then number of clusters is somewhat arbitrary, as I said above.

---

So what conclusions I draw from all of this.
- I'd like to keep the original "something certain and explicitly specified" idea of linker flavor, instead of inferring things from strings and other properties.
    - This means no `:arbitrary-string` part for `-Clinker-flavor`, maybe this second component can be added to `-Clinker` instead in the future, although I don't think it's necessary now, because `-Clink-arg=-fuse-ld=/mold/or/whatever` still exists.
- I'd like to keep the number of linker flavor "clusters" minimal under existing circumstances, and make the set `(2)` user-facing (both in `-Clinker-flavor` and in json target specs).
    E.g. the (non-legacy) accepted string values will be
    ```
    gnu
    gnu-lld
    gnu-cc
    gnu-lld-cc
    darwin
    darwin-lld
    darwin-cc
    darwin-lld-cc
    wasm-lld
    wasm-lld-cc
    unix
    unix-cc
    msvc
    msvc-lld
    em-cc
    bpf
    ptx
    ```
	instead of the current
    ```
    ld
    ld.lld
    gcc
    ld64.lld
    wasm-ld
    msvc
    lld-link
    l4-bender
    em
    bpf-linker
    ptx-linker
    ```
    (I tried to keep the naming consistent.)

**Comment by petrochenkov** [comment]
I'll review the code in this PR a bit later, I expect that we can mostly land it as is.

**Comment by petrochenkov** [comment]
Since the free form component is removed, `gcc:lld` can be renamed to `gcc-lld` (*) and added to `enum LinkerFlavor`.
`fn add_link_args` can then fill link args for this flavor if flags for `gcc` are filled.
`LinkerFlavorCli` is also no longer necessary, all flavors are well known.

(*) Although I'd prefer to use the scheme from https://github.com/rust-lang/rust/pull/96827#issuecomment-1208441595 eventually.

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 2]
```diff
@@ -2690,29 +2694,110 @@ fn get_apple_sdk_root(sdk_name: &str) -> Result<String, String> {
...
+
+        cmd.arg({
+            let mut arg = OsString::from("-B");
+            arg.push(gcc_ld_dir);
+            arg
+        });
+        cmd.arg(format!("-Wl,-rustc-lld-flavor={}", sess.target.lld_flavor.as_str()));
+    } else {
+        // Otherwise, we were asked to use `lld` but not `rust-lld`.
+        cmd.arg("-fuse-ld=lld");
```
Comment (petrochenkov) [inline compiler/rustc_codegen_ssa/src/back/link.rs:2801]: Maybe this PR should wait for https://github.com/rust-lang/rust/pull/100200, in addition to fixing the regression it basically makes the "self-contained" and "lld" aspects of `-Zgcc-ld` orthogonal and removes useless checks like
```rust
    // - checking the `lld-wrapper`s exist in the sysroot
```
above, that check something that is not necessarily true, and will prevent finding system lld instead.

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 3]
```diff
@@ -2690,29 +2694,110 @@ fn get_apple_sdk_root(sdk_name: &str) -> Result<String, String> {
...
+    // generated when using rust-lld as linker"
+    let instrument_coverage =
+        cg.instrument_coverage.is_some() && cg.instrument_coverage != Some(InstrumentCoverage::Off);
+    let generate_profile = cg.profile_generate.enabled();
+    if crt_objects_fallback && (instrument_coverage || generate_profile) {
+        sess.warn(
+            "Using `lld`, self-contained linking, and coverage or profile generation has known \
+            issues. See issue #79555 for more details, at \
+            https://github.com/rust-lang/rust/issues/79555",
+        );
```
[the remark this answers is not in the document]
Comment (petrochenkov) [reply compiler/rustc_codegen_ssa/src/back/link.rs:2766]: Same about other cases of copypasting issue tracker into the compiler's source code.

File: src/test/ui/linkers/gcc_ld_mismatch.rs [hunk 4]
```diff
@@ -0,0 +1,7 @@
+// build-fail
+// compile-flags: -Zgcc-ld=lld -Clinker-flavor=em -Zunstable-options
+
+// Test ensuring that until the unstable flag is removed (if ever), if both the linker-flavor and
+// `gcc-ld` flags are used, they ask for the same linker.
```
Comment (petrochenkov) [inline src/test/ui/linkers/gcc_ld_mismatch.rs:5]: If making these two options incompatible helps to remove at least some code or conditions in `fn handle_cli_linker_flavors` then please let's do that instead.

**Comment by lqd** [comment]
superseded by #112910

**Pull request closed without merging**